server/service/osquery.go

```diff
@@ -1405,3 +1411,139 @@ func (svc *Service) SubmitResultLogs(ctx context.Context, logs []json.RawMessage
+// Query Reports
+////////////////////////////////////////////////////////////////////////////////
+
+const (
+	maxQueryReportRows = 1000
+)
+
+func (svc *Service) SaveResultLogsToQueryReports(ctx context.Context, results []json.RawMessage) error {
+	// skipauth: Authorization is currently for user endpoints only.
+	svc.authz.SkipAuthorization(ctx)
+
+	// Do not insert results if query reports are disabled globally
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return newOsqueryError("getting app config: " + err.Error())
+	}
+	if appConfig.ServerSettings.QueryReportsDisabled {
+		return nil
+	}
+
+	var queryResults []fleet.ScheduledQueryResult
+	for _, raw := range results {
+		var result fleet.ScheduledQueryResult
+		if err := json.Unmarshal(raw, &result); err != nil {
+			return newOsqueryError("unmarshalling individual result log: " + err.Error())
+		}
+		queryResults = append(queryResults, result)
+	}
+
+	// Filter results to only the most recent for each query
+	filtered := getMostRecentResults(queryResults)
+
+	for _, result := range filtered {
+		if err := svc.processResults(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (svc *Service) processResults(ctx context.Context, result fleet.ScheduledQueryResult) error {
+	query, err := svc.ds.QueryByName(ctx, nil, getQueryNameFromResult(result.QueryName))
+	if err != nil {
+		return newOsqueryError("getting query by Name: " + err.Error())
+	}
+
+	// Discard Result if query is marked as discard data or if the query report is full
+	rowCount, err := svc.ds.ResultCountForQuery(ctx, query.ID)
+	if err != nil {
+		return newOsqueryError("getting result count for query: " + err.Error())
+	}
+	if query.DiscardData || rowCount >= maxQueryReportRows {
+		return nil
+	}
+
+	host, err := svc.ds.HostByIdentifier(ctx, result.OsqueryHostID)
+	if err != nil {
+		return newOsqueryError("getting host ID: " + err.Error())
+	}
+
+	// Delete any existing query results for host before inserting new results
+	err = svc.ds.DeleteQueryResultsForHost(ctx, host.ID, query.ID)
+	if err != nil {
+		return newOsqueryError("deleting query results for host: " + err.Error())
+	}
+
+	return svc.saveResultRows(ctx, result, query.ID, host.ID, rowCount)
+}
+
+// The "snapshot" array in a ScheduledQueryResult can contain multiple rows.  Each
+// row is saved as a separate ScheduledQueryResultRow. ie. a result could contain
+// many USB Devices or a result could contain all User Accounts on a host.
+func (svc *Service) saveResultRows(ctx context.Context, result fleet.ScheduledQueryResult, queryID, hostID uint, rowCount int) error {
+	fetchTime := time.Now()
+
+	for _, snapshotItem := range result.Snapshot {
+		if rowCount >= maxQueryReportRows {
+			break
+		}
+
+		row := &fleet.ScheduledQueryResultRow{
+			QueryID:     queryID,
+			HostID:      hostID,
+			Data:        snapshotItem,
+			LastFetched: fetchTime,
+		}
+
+		if _, err := svc.ds.SaveQueryResultRow(ctx, row); err != nil {
```

Comment:
If `svc.ds.SaveQueryResultRow` fails for some of the rows, then one issue that could happen is that we store some of the results but not all (because we are storing them one by one instead of in a transaction).